### ISSUE TITLE
train/test split for the g/h classifier ignores shower reuse, so the same showers end up on both sides

### ISSUE THREAD

**Issue by vuillaut** [cta-observatory/cta-lstchain#1429]
In `build_models` (`dl1_to_dl2.py`), the gamma sample is split into two parts to train the g/h classifier:

`train, testg = train_test_split(df_gamma, test_size=test_size, train_size=train_size)`

- train (default 80%) is used to train temporary energy and disp regressors.
- These temporary regressors reconstruct testg (default 20%) and the protons.
- The classifier is then trained on testg + protons, using the reconstructed features (log_reco_energy, reco_disp_norm, reco_disp_sign, signed_skewness, signed_time_gradient).

The two-step scheme exists so that the classifier sees gammas whose reconstructed features have realistic errors, the same errors they will have when the final models run on unseen data.

However,  MC data reuses each CORSIKA shower several times (10 times in the standard productions), each time with a different core position. 
Every copy of a shower has the same primary energy, direction and longitudinal development (Xmax, fluctuations). Only the impact point and the noise differ. Because the split ignores which shower an event comes from, copies of the same shower land on both sides.

Therefore, the classifier trains on gammas that are reconstructed too well. The temporary regressors have partly seen the `testg` showers, so the energy, disp norm and disp sign of the classifier's training gammas are better than they will be on independent data. 
This completely defeats the supposed advantage of splitting into intermediate training sets.

I have not measured the effect, which would require training models with a fix and comparing the performances.

We could either fix by shuffling the data properly to avoid reused shower events, but in my opinion the entire intermediate sets could be avoided altogether.
